File: solutions/2201-valid-arrangement-of-pairs/valid-arrangement-of-pairs.js
var validArrangement = function(pairs) {
    const adjacencyList = new Map();
    const inOutDegree = new Map();
    
    // Build graph and count in/out degrees
    for (const [from, to] of pairs) {
        if (!adjacencyList.has(from)) {
            adjacencyList.set(from, []);
        }
        adjacencyList.get(from).push(to);
        
        inOutDegree.set(from, (inOutDegree.get(from) || 0) + 1);
        inOutDegree.set(to, (inOutDegree.get(to) || 0) - 1);
    }
    
    // Find starting node
    let startNode = pairs[0][0];
    for (const [node, degree] of inOutDegree) {
        if (degree === 1) {
            startNode = node;
            break;
        }
    }
    
    const path = [];
    const nodeStack = [startNode];
    
    while (nodeStack.length > 0) {
        const neighbors = adjacencyList.get(nodeStack[nodeStack.length - 1]) || [];
        if (neighbors.length === 0) {
            path.push(nodeStack.pop());
        } else {
            const nextNode = neighbors.pop();
            nodeStack.push(nextNode);
        }
    }
    
    const arrangement = [];
    for (let i = path.length - 1; i > 0; i--) {
        arrangement.push([path[i], path[i-1]]);
    }
    
    return arrangement;
};

File: solutions/2201-valid-arrangement-of-pairs/valid-arrangement-of-pairs.rb
def valid_arrangement(pairs)
    adjacency_list = Hash.new { |h, k| h[k] = [] }
    in_out_degree = Hash.new(0)
    
    # Build graph and count in/out degrees
    pairs.each do |pair|
        adjacency_list[pair[0]] << pair[1]
        in_out_degree[pair[0]] += 1  # out-degree
        in_out_degree[pair[1]] -= 1  # in-degree
    end
    
    # Find starting node
    start_node = pairs[0][0]
    in_out_degree.each do |node, degree|
        if degree == 1
            start_node = node
            break
        end
    end
    
    path = []
    node_stack = [start_node]
    
    while !node_stack.empty?
        neighbors = adjacency_list[node_stack.last]
        if neighbors.empty?
            path << node_stack.pop
        else
            next_node = neighbors.pop
            node_stack << next_node
        end
    end
    
    arrangement = []
    path_size = path.size
    
    (path_size - 1).downto(1) do |i|
        arrangement << [path[i], path[i-1]]
    end
    
    arrangement
end